lib/srv/db/common/srvconnect/srvconnect.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	response, err := params.CertSigner.SignDatabaseCSR(ctx, &proto.DatabaseCSRRequest{
+		CSR:         csr,
+		ClusterName: params.Identity.RouteToCluster,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := keys.TLSCertificateForSigner(privateKey, response.Cert)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pool := x509.NewCertPool()
+	for _, caCert := range response.CACerts {
+		ok := pool.AppendCertsFromPEM(caCert)
+		if !ok {
+			return nil, trace.BadParameter("failed to append CA certificate")
+		}
+	}
+
+	return &tls.Config{
+		ServerName:   params.Server.GetHostname(),
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	}, nil
+}
+
+// ShuffleFunc defines a function that shuffles a list of database servers.
+type ShuffleFunc func([]types.DatabaseServer) []types.DatabaseServer
+
+// ShuffleSort is a ShuffleFunc that sorts database servers by name and host ID.
+// Used to provide predictable behavior in tests.
+func ShuffleSort(servers []types.DatabaseServer) []types.DatabaseServer {
+	sort.Sort(types.DatabaseServers(servers))
+	return servers
+}
+
+// ShuffleRandom is a ShuffleFunc that randomizes the order of database servers.
+// Used to provide load balancing behavior when proxying to multiple agents.
+func ShuffleRandom(servers []types.DatabaseServer) []types.DatabaseServer {
+	rand.New(rand.NewSource(time.Now().UnixNano())).Shuffle(
+		len(servers), func(i, j int) {
+			servers[i], servers[j] = servers[j], servers[i]
+		})
+	return servers
+}
+
+type Dialer interface {
+	// Dial dials any address within the site network, in terminating
+	// mode it uses local instance of forwarding server to terminate
+	// and record the connection.
+	Dial(params reversetunnelclient.DialParams) (conn net.Conn, err error)
+}
+
+// ConnectParams contains parameters for connecting to the database server.
+type ConnectParams struct {
+	Logger *slog.Logger
+	// Identity contains the identity information.
+	Identity tlsca.Identity
+	// Servers is the list of database servers that can handle the connection.
+	Servers []types.DatabaseServer
+	// ShuffleFunc is a function used to shuffle the list of database servers.
+	ShuffleFunc ShuffleFunc
+	// Cluster represents the cluster to which the database belongs.
+	Cluster Cluster
+	// Cluster represents the cluster to which the database belongs.
+	Dialer Dialer
+	// CertSigner is used to sign certificates for authenticating with the
+	// database.
+	CertSigner DatabaseCertificateSigner
+	// AuthPreference provides the authentication preferences for the cluster.
+	AuthPreference AuthPreferenceGetter
+	// ClientSrcAddr is the source address of the client making the connection.
+	ClientSrcAddr net.Addr
+	// ClientDstAddr is the destination address of the client making the
+	// connection.
+	ClientDstAddr net.Addr
+}
+
+// ConnectStats contains statistics about the connection attempts.
+type ConnectStats struct {
```

Comment:
nit: alternatively pass in an interface

Reply:
sorry what i meant is to pass an optional stats interface as part of param. it's ok this way so up to you.